### PR TITLE
fix: add more telemetry to apex replay debugger

### PR DESCRIPTION
**Pull request by CristiCanizales**
### What does this PR do?
- Adds more telemetry to the Apex Replay Debugger.

### What issues does this PR fix or reference?
@W-16499105@

### CODE REVIEW

File: packages/salesforcedx-apex-replay-debugger/src/adapter/apexReplayDebug.ts [hunk 1]
```diff
@@ -336,6 +370,12 @@ export class ApexReplayDebug extends LoggingDebugSession {
     this.printToDebugConsole(nls.localize('session_terminated_text'));
     response.success = true;
     this.sendResponse(response);
+    this.sendEvent(
+      new Event(SEND_METRIC_GENERAL_EVENT, {
+        subject: 'disconnectRequest',
+        type: 'apexReplayDebuggerDisconnect'
+      } as MetricGeneral)
```
Comment (peternhale) [inline packages/salesforcedx-apex-replay-debugger/src/adapter/apexReplayDebug.ts:377]: @CristiCanizales the `as` are not necessary, given Event.body is type `any`.

**Review by peternhale — COMMENTED** [review]
Looks good so far.

One question though. These changes are using the Event class to pass to sendEvent. Any particular reason? There are others available such OutputEvent.

File: packages/salesforcedx-apex-replay-debugger/src/adapter/apexReplayDebug.ts [hunk 2]
```diff
@@ -556,10 +627,14 @@ export class ApexReplayDebug extends LoggingDebugSession {
...
-            .get(uri)!
-            .push(this.convertClientLineToDebugger(bp.line));
+        if (!isVerified) {
+          this.sendEvent(
+            new Event(SEND_METRIC_ERROR_EVENT, {
+              subject: 'Failed to set breakpoint',
+              callstack: new Error().stack,
+              message: `Failed to set breakpoint at line ${bp.line} in ${args.source.path}`
+            })
+          );
```
Comment (daphne-sfdc) [inline packages/salesforcedx-apex-replay-debugger/src/adapter/apexReplayDebug.ts:637]: Why is this behavior removed?  Should it be converted to an if/else statement instead?

File: packages/salesforcedx-apex-replay-debugger/src/adapter/apexReplayDebug.ts [hunk 3]
```diff
@@ -22,9 +22,10 @@ import {
   Variable
 } from '@vscode/debugadapter';
 import { DebugProtocol } from '@vscode/debugprotocol';
-import { MetricError, MetricLaunch } from '..';
+import { MetricError, MetricGeneral, MetricLaunch } from '..';
```
Comment (daphne-sfdc) [inline packages/salesforcedx-apex-replay-debugger/src/adapter/apexReplayDebug.ts:25]: Unused imports - were you planning to use them for anything?

Comment (CristiCanizales) [reply packages/salesforcedx-apex-replay-debugger/src/adapter/apexReplayDebug.ts:25]: Oh no, I just missed to remove them in my previous commit.

File: packages/salesforcedx-apex-replay-debugger/src/adapter/apexReplayDebug.ts (hunk 2 above)
Comment (CristiCanizales) [reply packages/salesforcedx-apex-replay-debugger/src/adapter/apexReplayDebug.ts:637]: oops, deleted by mistake

**Comment by CristiCanizales** [comment]
> Looks good so far.
> 
> One question though. These changes are using the Event class to pass to sendEvent. Any particular reason? There are others available such OutputEvent.

I just wanted a general event 🙂

**Review by peternhale — APPROVED** [review]

File: packages/salesforcedx-apex-replay-debugger/test/unit/adapter/apexReplayDebug.test.ts
Comment (daphne-sfdc) [inline packages/salesforcedx-apex-replay-debugger/test/unit/adapter/apexReplayDebug.test.ts:1]: I see that `as MetricLaunch` is still used in Lines 172, 203, and 331.  I tried removing all 3 instances and the test still passes.  Can you remove it from those 3 lines and also remove the import statement in Line 19?

**Review by daphne-sfdc — APPROVED** [review]

File: packages/salesforcedx-vscode-apex-replay-debugger/src/telemetry/telemetry.ts [hunk 4]
```diff
@@ -73,6 +73,19 @@ export class TelemetryService {
     }
   }
 
+  public sendGeneralEvent(errorMsg: string, type: string, qty?: string): void {
+    if (this.reporters !== undefined && this.isTelemetryEnabled) {
+      this.reporters.forEach(reporter => {
+        reporter.sendTelemetryEvent('ardGeneralEvent', {
+          extensionName: EXTENSION_NAME,
+          errorMessage: errorMsg,
```
Comment (diyer) [inline packages/salesforcedx-vscode-apex-replay-debugger/src/telemetry/telemetry.ts:81]: is this event emitted for all events including successful events? the errorMessage field indicates these are only for errors, but would like to confirm

Comment (CristiCanizales) [reply packages/salesforcedx-vscode-apex-replay-debugger/src/telemetry/telemetry.ts]: Good catch! It was a typo, it's supposed to be the subject. I just pushed the change. Thanks!

Comment (diyer) [reply packages/salesforcedx-vscode-apex-replay-debugger/src/telemetry/telemetry.ts]: ty!

**Pull request merged**